snapshots/rawblock/rawblock.go

```diff
@@ -0,0 +1,570 @@
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/platforms"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/containerd/containerd/snapshots/storage"
+	"golang.org/x/sys/unix"
+
+	"github.com/containerd/continuity/fs"
+	"github.com/pkg/errors"
+)
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type:   plugin.SnapshotPlugin,
+		ID:     "rawblock",
+		Config: &SnapshotterConfig{},
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			ic.Meta.Platforms = append(ic.Meta.Platforms, platforms.DefaultSpec())
+
+			config, ok := ic.Config.(*SnapshotterConfig)
+			if !ok {
+				return nil, errors.New("invalid rawblock config")
+			}
+			config.setDefaults(ic.Root)
+
+			if err := config.validate(); err != nil {
+				return nil, errors.Wrap(err, "invalid rawblock config")
+			}
+
+			return NewSnapshotter(ic.Context, config)
+		},
+	})
+}
+
+type snapshotter struct {
+	root    string
+	sizeMB  uint32
+	fstype  string
+	options []string
+	ms      *storage.MetaStore
+}
+
+func testCopyFileReflink(source, target string) error {
+	src, err := os.Open(source)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open source %s", source)
+	}
+	defer src.Close()
+	tgt, err := os.Create(target)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open target %s", target)
+	}
+	defer tgt.Close()
+
+	st, err := src.Stat()
+	if err != nil {
+		return errors.Wrap(err, "unable to stat source")
+	}
+	_, err = unix.CopyFileRange(int(src.Fd()), nil, int(tgt.Fd()), nil, int(st.Size()), 0)
+
+	return err
+}
+
+func testReflinkCapability(dir string) (bool, error) {
+	file1 := filepath.Join(dir, "test-reflink-support-src")
+	file2 := filepath.Join(dir, "test-reflink-support-dst")
+	if out, err := exec.Command("truncate", "--size=1024", file1).CombinedOutput(); err != nil {
+		return false, errors.Errorf("%s:%v", string(out), err)
+	}
+	defer os.RemoveAll(file1)
+	defer os.RemoveAll(file2)
+	if err := testCopyFileReflink(file1, file2); err != nil {
+		return false, nil
+	}
+	return true, nil
+}
+
+// NewSnapshotter returns a Snapshotter which copies layers on the underlying
+// file system. A metadata file is stored under the root.
+//
+// Rawblock snapshot layout: each type of snapshots lives in its own directory.
+// root dir -> Active
+//          -> View
+//          -> Committed
+func NewSnapshotter(ctx context.Context, config *SnapshotterConfig) (snapshots.Snapshotter, error) {
+	if err := os.MkdirAll(config.RootPath, 0755); err != nil {
+		return nil, err
+	}
+
+	var (
+		active    = filepath.Join(config.RootPath, snapshots.KindActive.String())
+		view      = filepath.Join(config.RootPath, snapshots.KindView.String())
+		committed = filepath.Join(config.RootPath, snapshots.KindCommitted.String())
+	)
+
+	for _, path := range []string{
```

Comment:
The overlayfs and native drivers just use a single directory "snapshots". The cases where we would put them in a different directory are when there is a difference to the on disk representations (i.e. btrfs read only). If the snapshotted files are exactly the same then copy is not necessary. Note that is not a requirement for snapshotters to enforce no one still has the active mounted and is making changes. We are considering adding a new command in the future thought that would allow continue snapshotting of an active snapshot, but Commit here is not that.

Reply:
Thanks for the tips! What's your suggestion about it? Shall I keep active+view to be future proof or just use a single directory as overlayfs/native does?

Reply:
I think keeping them in a single directory is best since it is simpler and requires less file moving on commit. The necessary metadata about state is already held in the database so having that on the filesystem is redundant. Additionally, mount parameters are used to keep view/active different, storing them in separate directories doesn't make much of a different in terms of making one read only.

Reply:
Sorry for the late response! I totally agree with you and will update the PR as suggested.